src/app/dim-ui/Sheet.tsx

```diff
@@ -225,20 +184,60 @@ export default function Sheet({
         dragHandle.current?.contains(e.target as Node) ||
         sheetContents.current!.scrollTop === 0
       ) {
-        dragging.current = true;
+        dragControls.start(e);
       }
     },
-    []
+    [dragControls]
   );
 
-  const dragHandleUp = useCallback(() => (dragging.current = false), []);
+  // When drag ends we determine if the sheet should be closed either via the final
+  // drag velocity or if the sheet has been dragged halfway the down from its height.
+  const handleDragEnd = useCallback(
+    (_event: TouchEvent | MouseEvent | PointerEvent, info: PanInfo) => {
+      if (
+        info.velocity.y > dismissVelocity ||
+        (sheet.current && info.offset.y > dismissAmount * sheet.current.clientHeight)
+      ) {
+        triggerClose();
+        return;
+      }
+      animationControls.start('open');
+    },
+    [animationControls, triggerClose]
+  );
+
+  useLayoutEffect(() => {
+    if (freezeInitialHeight && sheetContents.current && !frozenHeight) {
+      if (sheetContents.current.clientHeight > 0) {
+        setFrozenHeight(sheetContents.current.clientHeight);
+      } else {
+        setTimeout(() => {
+          sheetContents.current && setFrozenHeight(sheetContents.current.clientHeight);
+        }, 500);
+      }
+    }
+  }, [freezeInitialHeight, frozenHeight]);
```

Comment:
No change, just moving things around to put state, callbacks, effects in that order. I find it makes it easier to reason about the component when things are in a heirarchy.